File: python/snake.py
#!/usr/bin/env python3
import json
import math
from http.server import HTTPServer, BaseHTTPRequestHandler

class RequestHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        handle_get_meta_data(self)

    def do_POST(self):
        if self.path == '/start':
            handle_game_started(self)
        elif self.path == '/move': 
            handle_move(self)

def handle_get_meta_data(request_handler):
    content = json.dumps({
        'apiversion': '1',
        'author': 'robvanderleek',
        'version': '1.0',
        'color': '#3776ab',
        'head': 'safe',
        'tail': 'sharp'
    })
    request_handler.send_response(200)
    request_handler.end_headers()
    request_handler.wfile.write(content.encode('utf-8'))

def handle_game_started(request_handler):
    body = get_body(request_handler)
    game_id = body['game']['id']
    print(f'Game started: {game_id}');
    request_handler.send_response(200)
    request_handler.end_headers()

def get_body(request_handler):
    content_length = int(request_handler.headers['Content-Length'])
    post_data = request_handler.rfile.read(content_length)
    return json.loads(post_data.decode('utf-8'))

def handle_move(request_handler):
    body = get_body(request_handler)
    turn = body['turn']
    print(f'Turn: {turn}')
    board = body['board']
    snake = body['you']
    direction = get_direction(board, snake)
    request_handler.send_response(200)
    request_handler.end_headers()
    request_handler.wfile.write(f'{{"move": "{direction}"}}'.encode('utf-8'))

def get_direction(board, snake):
    head = snake['head']
    directions = preferred_directions(board, head)
    return select_direction(board, head, directions)

def preferred_directions(board, head):
    food = nearest_food(head, board['food'])
    result = []
    if head['x'] != food['x']:
        result.append('right' if head['x'] < food['x'] else 'left')
    if head['y'] != food['y']:
        result.append('up' if head['y'] < food['y'] else 'down')
    for d in ['right', 'down', 'left', 'up']:
        if not d in result:
            result.append(d)
    return result

def nearest_food(head, food):
    food.sort(key=lambda f: distance(head, f))
    return food[0]

def distance(p1, p2):
    return math.sqrt(pow(p2['x'] - p1['x'], 2) + pow(p2['y'] - p1['y'], 2))

def select_direction(board, head, directions):
    for d in directions:
        if d == 'left' and free_cell(board, head['x'] - 1, head['y']):
            return 'left'
        if d == 'right' and free_cell(board, head['x'] + 1, head['y']):
            return 'right'
        if d == 'down' and free_cell(board, head['x'], head['y'] - 1):
            return 'down'
        if d == 'up' and free_cell(board, head['x'], head['y'] + 1):
            return 'up'
    else:
        print('Oops')

def free_cell(board, x, y):
    for s in board['snakes']:
        if {'x': x, 'y': y} in s['body']:
            return False
    return x >= 0 and y >= 0 and x < board['width'] and y < board['height']

PORT=3001

print(f'Starting Battlesnake server on port: {PORT}')
server = HTTPServer(('0.0.0.0', PORT), RequestHandler)
server.serve_forever()

File: javascript/snake.js
#!/usr/bin/env node
const http = require('http');

const requestHandler = (req, res) => {
    if (req.url === '/') {
        handleGetMetaData(req, res);
    } else if (req.url === '/start') {
        handleWithBody(req, res, handleGameStart);
    } else if (req.url === '/move') {
        handleWithBody(req, res, handleMove);
    }
}

const handleGetMetaData = (req, res) => {
    const response = {
        'apiversion': '1',
        'author': 'robvanderleek',
        'color': '#C00000',
        'head': 'caffeine',
        'tail': 'curled',
        'version': '1.0'
    }
    res.end(JSON.stringify(response));
}

const handleWithBody = (req, res, handler) => {
    let body = [];
    req.on('data', (chunk) => {
        body.push(chunk);
    }).on('end', () => {
        body = Buffer.concat(body).toString();
        handler(req, res, JSON.parse(body));
    });
}

const handleGameStart = (req, res, body) => {
    console.log(`Game started: ${body.game.id}`);
    res.writeHead(200);
}

const handleMove = (req, res, body) => {
    const {turn, you: snake, board} = body;
    console.log(`/move called for turn: ${turn}`);
    const dir = getDirection(board, snake);
    res.setHeader('Content-Type', 'application/json');
    res.writeHead(200);
    res.end(`{"move": "${dir}"}`);
}

const getDirection = (board, snake) => {
    const directions = preferredDirections(board, snake.head)
    return selectDirection(board, snake.head, directions)
}

const preferredDirections = (board, head) => {
    const food = nearestFood(head, board.food);
    let result = []
    if (head.y != food.y) {
        result.push(head.y < food.y ? 'up' : 'down');
    }
    if (head.x != food.x) {
        result.push(head.x < food.x ? 'right' : 'left');
    }
    return result.concat(['left', 'up', 'right', 'down']
        .filter(d => !result.includes(d)));
}

const nearestFood = (head, food) => food.sort((f1, f2) => distance(head, f1) - distance(head, f2))[0]

const distance = (p1, p2) => Math.sqrt(Math.pow(p2.x - p1.x, 2) + Math.pow(p2.y - p1.y, 2));

const selectDirection = (board, head, directions) => {
    for (const d of directions) {
        if (d === 'left' && freeCell(board, {x: head.x - 1, y: head.y})) {
            return 'left'
        } else if (d === 'right' && 
                freeCell(board, {x: head.x + 1, y: head.y})) {
            return 'right'
        } else if (d === 'down' && 
                freeCell(board, {x: head.x, y: head.y - 1})) {
            return 'down'
        } else if (d === 'up' && freeCell(board, {x: head.x, y: head.y + 1})) {
            return 'up'
        }
    }
    console.log('Oops');
}

const freeCell = (board, c) => {
    if (board.snakes.flatMap(s => s.body).some(p => equal(c, p))) {
        return false;
    }
    return c.x >= 0 && c.y >= 0 && c.x < board.width && c.y < board.height;
}

const equal = (p1, p2) => p1.x === p2.x && p1.y === p2.y;

const PORT = 3000;
const server = http.createServer(requestHandler);
server.listen(PORT, '0.0.0.0', () => console.log(`Battlesnake server running on port: ${PORT}`));
